module/dal/test/src/test/java/org/openbase/bco/dal/test/action/RemoteActionTest.kt

```diff
@@ -0,0 +1,288 @@
+        Assertions.assertFalse(
+            dominantActionExtentionFlag.value,
+            "Dominant action was extended even through the action was already canceled."
+        )
+        println("cancel low prio action")
+        lowPrioLongtermAction.cancel().get()
+        lowPrioLongtermActionExtentionFlag.value = false
+        println("validate everything is done")
+        for (unit in units) {
+            unit.requestData().get()
+            for (actionDescription in unit.actionList) {
+
+                // filter termination action
+                if (actionDescription.priority == ActionPriorityType.ActionPriority.Priority.TERMINATION) {
+                    continue
+                }
+                Assertions.assertTrue(
+                    RemoteAction(actionDescription).isDone,
+                    "Zombie[" + actionDescription.actionState.value.name + "] actions detected: " + MultiLanguageTextProcessor.getBestMatch(
+                        actionDescription.description
+                    )
+                )
+            }
+        }
+        println("wait until last extension timeout and validate if no further extension will be performed for low prio longterm action...")
+        Thread.sleep(lowPrioLongtermAction.getValidityTime(TimeUnit.MILLISECONDS) + 10)
+        Assertions.assertEquals(
+            false,
+            lowPrioLongtermActionExtentionFlag.value,
+            "Low prio action was extended even through the action was already canceled."
+        )
+        println("validate if still everything is done")
+        for (unit in units) {
+            for (actionDescription in unit.actionList) {
+
+                // filter termination action
+                if (actionDescription.priority == ActionPriorityType.ActionPriority.Priority.TERMINATION) {
+                    continue
+                }
+                Assertions.assertEquals(
+                    true,
+                    RemoteAction(actionDescription).isDone,
+                    "Zombie[" + actionDescription.actionState.value.name + "] actions detected: " + MultiLanguageTextProcessor.getBestMatch(
+                        actionDescription.description
+                    )
+                )
+            }
+        }
+        println("test successful")
+    }
+
+    @Test
+    fun `should not fail on cancel if action was never executed`() {
+
+        val failedServiceCall = CompletableFuture.failedFuture<ActionDescription>(CancellationException())
+        RemoteAction(failedServiceCall).cancel().get()
+    }
+
+    internal data class Flag(
+        @Volatile
+        var value: Boolean = false,
+    )
+
+    companion object {
+        private val LOGGER = LoggerFactory.getLogger(RemoteActionTest::class.java)
+        private var mrPinkUserToken: AuthTokenType.AuthToken? = null
+        private var mrPinkUserId: String? = null
+        private var mrPinkActionParameter: ActionParameterType.ActionParameter? = null
+    }
```

Comment:
Is it normal that these are defined at the bottom of the file or would you mind moving them to the top?

Reply:
moved up